### DISCUSSION TITLE
Deserialization of nested SQLModel(s) not working when 'table=True'

### DISCUSSION THREAD

**Discussion by emilhe** [fastapi/sqlmodel#961] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Literal, Union

from sqlalchemy import JSON, Column
from sqlmodel import Field, SQLModel


class Cat(SQLModel):
    type: Literal["cat"] = "cat"
    calico: bool


class Dog(SQLModel):
    type: Literal["dog"] = "dog"
    service_animal: bool


Animal = Union[Cat, Dog]


class Pet(SQLModel, table=True):
    name: str = Field(primary_key=True)
    animal: Animal = Field(discriminator="type", sa_column=Column(JSON))


jiji = Pet(name="Jiji", animal=Cat(calico=True))
print(jiji.model_dump_json())  #  {"name":"Jiji","animal":{"type":"cat","calico":true}}
jiji_copy = Pet.model_validate_json(jiji.model_dump_json())
print(jiji)  #  name='Jiji' animal=Cat(type='cat', calico=True)
print(jiji_copy)  # name='Jiji' animal={'type': 'cat', 'calico': True}
```

### Description

I am trying to bind nested models to a JSON field. The serialization/deserialization process works as expected when `table=False`, but when i set `table=True`, the deserialization is no longer working as intended. In both cases, the serialized form is,

```
{"name":"Jiji","animal":{"type":"cat","calico":true}}
```

When `table=False` the deserialization yields,

```
name='Jiji' animal=Cat(type='cat', calico=True)
```

as exptected. However, when i set `table=True`, the nested model (here, the `animal ` field) ends up as a `dict ` instead of an object of the intended type (here, `Cat`),

```
name='Jiji' animal={'type': 'cat', 'calico': True}
```

I found some related discussions, https://github.com/tiangolo/sqlmodel/issues/63 and https://github.com/tiangolo/sqlmodel/issues/52, that indicates that this behavior may be related to issues with pydantic model validation. 

Now, my question is - is there any (proper) way to mitigate this issue?

### Operating System

Linux

### SQLModel Version

0.0.18

### Python Version

Python 3.12.0

**Comment by dadodimauro** [comment]
Hi @emilhe, I'am encountering the same problem as you described. Where you able to solve this?

**Reply by emilhe** [reply]
I didn't find any proper solution.

**Comment by 20-buck-spin** [comment]
I am facing the same issue. For now I just model_validate the field when I intend to use it. But this is a very error prone hard to maintain approach. Any help would be highly appreciated.

My hunch is that I need to pass a special function to the json_deserializer param of create_engine. I am looking into it. I hope this helps someone find the necessary solution

**Comment by martinlatrille** [comment]
I have the same issue as well, thank you @20-buck-spin for the temporary workaround

**Comment by vsxh2001** [comment]
Got the same issue, I tried to use fastapi with sqlmodel that inherited from another sqlmodel, and the automatic json to pydantic didn't work properly (specifically a datetime object was interpreted as a string). thanks @20-buck-spin for the work around.

**Comment by regunakyle** [comment]
Bump, just got hit by the same issue. 

As a workaround, I use SQLAlchemy's reconstructor.

Full example:

```python
import datetime as dt
import typing as ty

from pydantic import BaseModel, Field, TypeAdapter
from sqlalchemy.orm import reconstructor
from sqlmodel import JSON, Float, LargeBinary, SQLModel, Unicode
from sqlmodel import Field as sqlField


class CronTriggerDTO(BaseModel):
    trigger_type: ty.Literal["cron"] = Field(default="cron")
    month: None | int = Field(default=None, ge=1, le=12)
    day: None | int = Field(default=None, ge=1, le=31)
    day_of_week: None | int = Field(default=None, ge=0, le=6)
    hour: None | int = Field(default=None, ge=0, le=23)
    minute: None | int = Field(default=None, ge=0, le=59)


class DateTriggerDTO(BaseModel):
    trigger_type: ty.Literal["date"] = Field(default="date")
    run_date: dt.datetime


TriggerDTO = ty.Annotated[
    CronTriggerDTO | DateTriggerDTO,
    Field(discriminator="trigger_type"),
]


class ScheduledJob(SQLModel, table=True):
    id: str = sqlField(primary_key=True, sa_type=Unicode(191))
    next_run_time: float = sqlField(sa_type=Float(25), index=True)
    job_state: bytes = sqlField(
        sa_type=LargeBinary,
        exclude=True,
    )
    description: str = sqlField(sa_type=Unicode(500), default="")
    trigger_dto_json: ty.Optional[TriggerDTO] = sqlField(
        sa_type=JSON, discriminator="trigger_type"
    )
    created_at: dt.datetime = sqlField(default_factory=lambda: dt.datetime.now(dt.UTC))

    @reconstructor
    def reconstructor(self) -> None:
        # https://github.com/fastapi/sqlmodel/discussions/961
        if isinstance(self.trigger_dto_json, dict):
            # Maybe add some try catch?
            self.trigger_dto_json = TypeAdapter(TriggerDTO).validate_python(
                self.trigger_dto_json
            )
```

**Comment by vawsgit** [comment]
See this: https://sqlmodel.tiangolo.com/tutorial/fastapi/relationships/
